### DISCUSSION TITLE
Fake weather provider for Gadgetbridge to display BG values on Smartwatches

### DISCUSSION THREAD

**Discussion by SaschaH42** [NightscoutFoundation/xDrip#2125] [Ideas]
[XdripGBGateway_2022-02-18_AndroidStudio.zip](https://github.com/NightscoutFoundation/xDrip/files/8695713/XdripGBGateway_2022-02-18_AndroidStudio.zip)
[apk File.zip](https://github.com/NightscoutFoundation/xDrip/files/8695717/apk.File.zip)
[Wattchface Amazfit GTR 47.zip](https://github.com/NightscoutFoundation/xDrip/files/8695725/Wattchface.Amazfit.GTR.47.zip)

Hello xDrip Team,
first of all I want to say a big THANK YOU for this really great piece of Software. My daughter has now T1D for one year and this Project is the largest support for having this more or less under control. You're doing a GREAT job !!!

Maybe with this post I can support the community a bit with a feature enhancement idea.
I'm wearing a Amazfit Bip or Amazfit GTR Smartwatch and use Gadgetbridge as Host Application on the Phone. I'd like to see the BG values on my smartwach. For sure xDrip supports some smartwatches - which is great. My daughter wears a Mi Band 4.
My idea is to use a Fake Weather provider for Gadgetbridge to display BG values instead of weather data. 
I did a small prove of concept - and I can tell it does work ! You can find all Files including the source code attached.
As an excuse upfront - I'm not a programmer nor very exepericenced.. So please bear this in mind when you have a look at the code. It's just a prove of concept...
I haven't managed to compile the xDrip code within Android Studio, so for the test I created an application with contacts the xDRIP web service and extracts the BG values and then mimics a weather service for Gadgetbride. Gadedbride then believes there is new weather data and uploads the new data as upper and lower temperature to the smartwatch.
It's working most of the day - but sometimes my phone kills my gateway app. I assume because of energy settings or some issues in my code.
If you would like to test this - just install the attached debug apk and start it. It'll immediately contact xdrip every 20s and uploads the "weather" data to gadgetbridge.
I've also created some watchfaces for the GTR to show the data a bit more nice.

Currently there are some shortfalls as well...
as the BG value is combines out of upper and lower temp (there is no e.g. 217C) it's using 21 and 7 then. 21C upper temp and 7C lower temp...
Based on the fact that the upper temp always needs to be higher than the lower temp this leads in some cases to wrong displays. e.g. falling and 107... this ends up in -110 ... I haven't found the time for a better calculation... But I really wanted to share the idea.

If there is somebody out there who can have a look at the code and would be able to add this feature to the xDrip project that would be awesome.

As stated before: I've tested this with an Amazfit Bip and GTR 47. But I'd assime this should work with all smartwatches which are supported by Gadgetbridge.

Thansk for reading all of this. 
You can find attached:
- my prove of concept apk file (don't miss to installl gadgetbrigde - e.g. from f-droid)
- the source code of the apk file
- 2 Watchfaces for the Amazfit GTR 47

I'm curious to hear what you thing about this idea.
Thanks.

**Comment by Nb-tst** [comment]
Great idea.

Tested "working" on a Mi Band 5. 
A xdrip-gadgetbridge integration would be very useful, more smartwatches supported and a lighter communication. Currently I am using Artem's version of xdrip to communicate with my band, works fine ( huge thanks to him ) but has some downsides, given that it uploads a new watchface every 5 minutes and it takes even 30secs syncing, draining a lot of battery and breaking communication with the app which I use to manage the band ( which is exactly Gadgetbridge in my case ). Having all communication managed by the same app would result for sure in a better stability and a lot less data transmitted.

Some ideas, idk if useful:
- maybe switching from celsius to fahrenheit is the best way to have a third digit for bg values
- weather animations ( eg sun/clouds ) should be replaced with the arrows indicating trends. I think this is possible, not sure at what level it has to be done.
- maybe 20sec is too frequent and it could impact on battery, a typical libre2-oop setting would give new readings every 5 minutes, so a constant updating is not useful
- have you tried to exclude the app from battery optimization in order to have a more stable behavior?

BTW, have you tried to post this idea in gadgetbridge's codeberg asking for a possible support in the official app? I think it could be a good thing even to that community to increase their user base.

**Comment by Der-Schubi** [comment]
Just want to point out another possibility: https://bigdigital.home.blog/2022/06/16/watchdrip-a-new-application-for-xdrip-watch-integration/

**Comment by Nb-tst** [comment]
Seems that here they're using the same approach:

https://insulinclub.de/index.php?thread/33795-free-three-ein-xposed-lsposed-modul-f%C3%BCr-libre-3-aktueller-wert-am-sperrbildschir/&postID=657413#post657413

( see gadgetbridge section of the comment )

**Comment by atomphil** [comment]
The Gadgetbridge feature was fortunately integrated into [Free Three](https://insulinclub.de/index.php?thread/33795-free-three-ein-xposed-lsposed-modul-f%C3%BCr-libre-3-aktueller-wert-am-sperrbildschir/) by @DecentWoodpecker67 .

Thanks to [the new editor](https://tnxec2.github.io/amazfit-bips-watchface-static/index.html), more values could now be displayed on the watchface, to be able to use these variables the Gadgetbridge connection of Free Three has been extended. Unfortunately the implementation of the weather data processing in Gadgetbridge is not complete and some of the possible values are not even changeable via the weather interface. **I have [opened an issue at Gadgetbridge](https://codeberg.org/Freeyourgadget/Gadgetbridge/issues/2949). Feel free to join or ask there and show your interest in the topic.**

**Comment by Nb-tst** [comment]
Tried to send some fake values to an Amazfit5  using @SaschaH42 application by playing with GadgetbridgeAPI.java file in its source code.
Current findings:

temperatures have actually a -128,127 range and there is no check, so you could send a maxTemp value which is lower than a minTemp, and the same applies to current temp
all you have to do is adding 17 to the value to display. Negative values are displayed from 128+17=145 onwards, in a descending sorting. So 148 is 148-17 = -125 and 248 is -25.
Humidity has a 0-255 range, so it could be potentially used for BG with some caveats ( eg 255 transformed to 000 as a conventional high value signal )
I still cannot correctly handle wind speed, but it should at least have a 0-10 range.

So there is plenty of digits to play with on an appropriate watchface.

I don't know if the same applies to the Amazfit 5 twin, aka Mi band 5, and his successor mi band 6.

**Comment by Navid200** [comment]
I don't understand the purpose of this entire thread.  I read the very first post until I got to Mi Band 4.
MiBand 4 is supported by WatchDrip.
Please explain why you don't want to use WatchDrip.

**Reply by Nb-tst** [reply]
As I stated in an earlier post, I am currently using watchdrip.

Pros of this new possible approach are multiple:
- No need to upload a watchface every 5 minutes as with watchdrip, so to reduce battery drain, increase range ( BLE could switch to lowspeed/increased range mode ), add the possibility of a more frequent update. Also using just one app to manage the smartwatch/band avoids conflicts and improve the connection/reconnection stability.
- It could potentially extend the possibility to display xDrip values to all watches supported by GadgetBridge
- This probably goes beyond xDrip primary purpose, but it would add another method to use a total privacy friendly application to manage BG data.

**Reply by Navid200** [reply]
You can contact WatchDrip and suggest to enhance it to add the features that you feel Watchdrip is currently missing.
Have you done that?

**Reply by atomphil** [reply]
> You can contact WatchDrip and suggest to enhance it to add the features that you feel Watchdrip is currently missing. Have you done that?

https://github.com/bigdigital/watchdrip/issues/6#

**Comment by Nb-tst** [comment]
An update: I managed to send xDrip values to an Amazfit Band 5 using the gadgetbridge weather method using a @SaschaH42  patched app.
This was the very first time I used Java, so code is probably very bad. 
Details [here](https://github.com/Nb-tst/glucose-to-weather), feel free to open a discussion there if here is not appropriate.

**Comment by 0x05a3b7c1d2e4f89fff** [comment]


**Reply by Navid200** [reply]
Have you communicated with the WatchDrip developer to see if you can help improve it?
It would be great to have collaboration rather than competition. 
If collaboration does not work, and there is no choice, it will be OK then.

Please don't use this repository for promoting an app before attempting to improve existing protocols.
Thanks

**Comment by Navid200** [comment]
Would you guys please not post links to apps you have created?
If you have an idea, the first step is to attempt to collaborate.  There is no reason for competition in the open source community.

Of course there are times an individual will refuse to collaborate or is not interested any longer in supporting their app.  If that is the case, it will become clear when you approach them.
But, please don't use competition as a first step.

There is a guideline in these discussions not to post links to other repositories unless it is absolutely necessary.

Please do not link to an app you have created before attempting to contact the existing app developers.
If you have questions, you can always open a discussion and ask and we can talk.  

Please remove all links from this thread or I will have to remove them.
Thanks
